Add maintainVisibleContentPosition support on Android

janicduplessis opened the pull request:
## Summary

This adds support for `maintainVisibleContentPosition` on Android. The implementation is heavily inspired from iOS, it works by finding the first visible view and its frame before views are update, then adjusting the scroll position once the views are updated.

Most of the logic is abstracted away in MaintainVisibleScrollPositionHelper to be used in both vertical and horizontal scrollview implementations.

Note that this only works for the old architecture, I have a follow up ready to add fabric support.

## Changelog

<!-- Help reviewers and the release process by writing your own changelog entry. For an example, see:
https://reactnative.dev/contributing/changelogs-in-pull-requests
-->

[Android] [Added] - Add maintainVisibleContentPosition support on Android

## Test Plan

Test in RN tester example on Android

https://user-images.githubusercontent.com/2677334/197319855-d81ced33-a80b-495f-a688-4106fc699f3c.mov

ryancat reviewed (commented):
Please take a look at the comment and make sure there's no race possible here.

janicduplessis commented on ReactAndroid/src/main/java/com/facebook/react/views/scroll/MaintainVisibleScrollPositionHelper.java:
```diff
@@ -0,0 +1,186 @@
+      if (deltaX != 0) {
+        int scrollX = mScrollView.getScrollX();
+        mScrollView.scrollTo(scrollX + deltaX, mScrollView.getScrollY());
+        mPrevFirstVisibleFrame = newFrame;
```
This is to avoid adjusting the scroll position multiple times. I had cases where this method would be called multiple times in a row.

janicduplessis commented on ReactAndroid/src/main/java/com/facebook/react/views/scroll/MaintainVisibleScrollPositionHelper.java:
```diff
@@ -0,0 +1,186 @@
+      return;
+    }
+
+    View firstVisibleView = mFirstVisibleView.get();
```
If willDispatchViewUpdates is called from ui thread (which I’m not sure if it ever does) the dispatch to ui thread will be done synchronously by runOnUIThread. Ideally we could have a way to observe before / after views are updated on the ui thread but wanted to avoid adding more methods to UIManagerObserver for old arch. In my tests this always worked so I was fine with this solution.

Rays15 replied:
@janicduplessis `firstVisibleView` may be null, i ran into.
`
E/unknown:ReactNative: Exception in native call
    java.lang.NullPointerException: Attempt to invoke virtual method 'void android.view.View.getHitRect(android.graphics.Rect)' on a null object reference
        at com.facebook.react.views.scroll.MaintainVisibleScrollPositionHelper.updateScrollPositionInternal(MaintainVisibleScrollPositionHelper.java:107)
        at com.facebook.react.views.scroll.MaintainVisibleScrollPositionHelper.updateScrollPosition(MaintainVisibleScrollPositionHelper.java:97)
        at com.facebook.react.views.scroll.ReactScrollView.onLayoutChange(ReactScrollView.java:1128)
        at android.view.View.layout(View.java:20717)
        at android.view.ViewGroup.layout(ViewGroup.java:6198)
        at com.facebook.react.uimanager.NativeViewHierarchyManager.updateLayout(NativeViewHierarchyManager.java:254)
        at com.facebook.react.uimanager.NativeViewHierarchyManager.updateLayout(NativeViewHierarchyManager.java:222)
`

janicduplessis replied:
I see, yes it is missing a null check, I'll open a PR to fix it.

janicduplessis commented on ReactAndroid/src/main/java/com/facebook/react/views/scroll/MaintainVisibleScrollPositionHelper.java:
```diff
@@ -0,0 +1,186 @@
+    Rect newFrame = new Rect();
+    firstVisibleView.getHitRect(newFrame);
+
+    if (mHorizontal) {
```
Do you have an example of the interface you have in mind?

My thinking when organizing the code like this was to reduce duplication between the 2 implementations while also moving all code related to maintain visible content position out of the scrollview classes to avoid making them too big. The common logic when the variant specific is also very intertwined in some cases, see https://github.com/facebook/react-native/pull/35049/files/c069ae3e39595e7d7e7a3259c6004d1313ad6ba8#diff-b0c8b7c192810637724d9d2ed35db2b8921c45e153ebbda1ef8f34cb53837473R153 for example.

aweffr commented:
any progress on this?

janicduplessis commented:
@skinsshark @ryancat Any updates on this? I think the only unadressed point is https://github.com/facebook/react-native/pull/35049#discussion_r1009966255.

areebbajwa commented:
Need an update on this please.

paula-morales commented:
Any update? 🙏

codal-mpawar commented:
ANy update on this?

skinsshark commented:
hey all! thank you for your patience and sorry for the delay, there are a couple things we want to test internally but haven't had the bandwidth just yet- i will get this prioritized and be back with updates! happy new year!

siu666 commented:
@pull-bot could yours rebuild the tarball in CircleCI ?, cause  it had been removed  after 30days  , thx~~~

matt-alice commented:
Hi, many happy returns and thanks for the update.
Is there any rough estimate RE: when this will be released? could it happen this month or next month?
Many thanks for your help.

areebbajwa commented:
I would like to know that as well

conoremclaughlin commented:
Would love to have this as well along with documentation with an example on how to set the parameters. Thank you so much!

skinsshark commented:
hi! i'll follow up again, but i expect to finish this for tomorrow. thanks all for your patience =)

skinsshark commented:
(ah! thought i clicked comment on friday....) hey @janicduplessis, thanks for this change! this looks good to go but i'm going to wait to land this on Monday so i can be online when it ships!

hadnet commented:
So this maintanVisibleContentPosition props is now available on what version of RN? And is available on FlatList as well, right?

janicduplessis commented:
It is available in 0.72. It also works with FlatList but there are a few known issues that will be fixed in https://github.com/facebook/react-native/pull/35993 which isn’t merged yet.

hantrungkien commented:
@janicduplessis  I have tried it in 0.72-rc06. Everything seems fine but I found a problem with my chat app with scenario:

1. In first request I load message from 30 to 0 and show it on FlatList with reverse.
2. In second request I load message from 35-31 and append to leading page.

If all messages are short content then maintainVisibleContentPosition is working fine. But messages from 35-31 is very long text (greater than 3000 char) then FlatList still auto scroll to first position.

Do you have any suggestions for me?

saif-o99 commented:
@janicduplessis Hey, the props maintanVisibleContentPosition is not exist in the whole react native repo, is it removed now?

haileyok commented:
@hantrungkien did you ever experiment further with that? thanks!

janicduplessis commented:
We also have noticed some issues on android, I will look at it soon.

haileyok commented:
@janicduplessis Following up in case this is helpful:

https://snack.expo.dev/@haileyok/flatlist-repro

```
/**
 * To repro:
 * - On initial render, we can add 10 or 20 items to the list without issue while maxToRenderPerBatch is set to 20.
 * - If we increase this number to 30, we will get a jump in position after adding the items above.
 * - If we then increase the maxToRenderPerBatch to 30, we can add 30 items above without issue.
 * 
 * Notice that the maxToRenderPerBatch seems to need to be set to >= the number of items we are adding.
 * 
 * Notably, after resetting the list, the issue isn't present anymore. It is only on the initial render that we see problems.
 */
 ```

Video: https://streamable.com/almoal

janicduplessis commented:
Thanks, this will help, planning to look at it this week.

janicduplessis commented:
No need for an issue.

perunt commented:
@haileyok, have you tried adding half the number of initial items? Also, as you mentioned, the `maxToRenderPerBatch` should not be less than the number of items you're adding.

haileyok commented:
> @haileyok, have you tried adding half the number of initial items? Also, as you mentioned, the `maxToRenderPerBatch` should not be less than the number of items you're adding.

my experience was that the number of initial items didn't affect anything. IIRC, even with just one initial item (and some padding on the `contentContainerStyle` to account for the needed space to maintain that position) we would still get a jump. *If* the item at index 0 (or I presume whatever the min index is set to) is the same height as the items being prepended though, there is never a jump regardless of any of the configuration.

I suppose my immediate question would be is this something that's a bug/could be fixed or is this a limit of the virtualization (I don't know all the specifics on how the virtualization works, just a quick review of the code here)?

I'm implementing a workaround right now to artificially "load in" items when we scroll up, which is a fine solution (and maybe better anyway for performance so we don't immediately load a bunch of items) but would still be interested in knowing if this could be fixed.

Appreciate it!

janicduplessis commented:
For more context:

Usually the problem with maintainVisibleContentPosition and virtualization is that `maintainVisibleContentPosition` relies on tracking the first visible view to maintain its position when items are added. Then what can happen is that the first visible view is virtualized away and replaced with a spacer if too many items are added and the old first visible view is pushed out of the virtualization window.

We added some mitigations to try and prevent this from happening, but for some reason it seems to still happens some times on Android.

janicduplessis commented:
@haileyok  I just realized the issue you mentionned happens on iOS. It actually looks like an issue I addressed recently. Sadly I don't think the changes haven't made it into a release yet.

Can you try to apply the changes from this https://github.com/facebook/react-native/commit/3ed4bf9046ae125d244283c3556892345bfd77e1. This code should be in the @react-native/virtualized-lists package.

janicduplessis commented:
The issue we are currently experiencing seems to happen only on Android if adding items at the start of the list during scrolling.

haileyok commented:
Ah, awesome! I'll look later if I have time today or tomorrow if not and let you know. Appreciate the quick response. I take it this addresses both platforms since it's just the JS side?

janicduplessis commented:
@haileyok It does yes

janicduplessis commented:
There is also one more case of sliding that I know of. It happens when items size vary a lot. Internally VirtualizedList uses the average size of previously measured items to estimate the size of new items being added before they get measured and the actual size can be used.

In your repro if I increase number of item added to 60 it starts happening.

miafoo commented:
> There is also one more case of sliding that I know of. It happens when items size vary a lot. Internally VirtualizedList uses the average size of previously measured items to estimate the size of new items being added before they get measured and the actual size can be used.

I'm running into this as well unfortunately. Do you have an issue or PR tracking this, or will you keep us posted here? Just want to stay updated for when it's fixed 💜

haileyok commented:
@miafoo Can't give you any updates on that status, but what I can pass over is a bit of how we worked around the problem.

One way that you might be able to get around this though is to only add items to the top of the list in increments. If you use `onStartReached`, you can achieve a bit of a "loading" effect, even if you already have all the items and don't have to fetch them. We did this here https://github.com/bluesky-social/social-app/blob/58aaad704aa971c5ebbf5a5f330a2e2129b557f6/src/view/com/post-thread/PostThread.tsx#L280.

janicduplessis commented:
@haileyok I opened a PR to fix another issue on Android https://github.com/facebook/react-native/pull/43425

wqhui commented:
awesome! Can 70.0.x still be supported? The react-native upgrade version  is too difficult...
